New page: placeholder attr description

Arisch24 opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

The placeholder attribute is fundamental for input and textarea types in html form elements.

This PR provides documentation for the placeholder attribute and its use cases with relevant examples.

### Motivation

To help readers understand placeholder attribute better and saving readers time as well.

### Additional details

[W3 placeholder attr](https://www.w3.org/TR/2011/WD-html5-author-20110809/common-input-element-attributes.html#the-placeholder-attribute)
[W3 schools placeholder attr](https://www.w3schools.com/tags/att_input_placeholder.asp)

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

Relates to #28375
<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

Arisch24 committed:
New page: placeholder attr description

estelle commented on files/en-us/web/html/attributes/placeholder/index.md:
```diff
@@ -0,0 +1,52 @@
+
+{{HTMLSidebar}}
+
+The **`placeholder`** attribute specifies a short hint ,a word or phrase, that describes the expected value of an input. A hint could be a sample value or a short description containing the format of the expected value. The attribute, if specified, must have a value that contains no U+000A LINE FEED (LF) or U+000D CARRIAGE RETURN (CR) characters.
```
Maybe something like this:

```
The **`placeholder`** defines the text displayed in a form control when the control has no value. The placeholder text provides a brief hint to the user as to the expected type of data that should be entered into the control.

The placeholder should be a word or short phrase that hints at the expected data type, not an explanation or prompt. The placeholder should not be used instead of a {{HTMLElement("label")}}. As the placeholder is not visible if the value of the form control is not null, using `placeholder` instead of a `<label>` for a prompt harms usability and accessibility.
```

and then moving the whitespace information to a note. the parts about it only appearing when there is no value and disappearing when there is, and not using it to replace a label are more important than the LF or CR. The LF/CR note should not be a copy paste from the spec. instead, explain what happens (the placeholder value gets clipped)

estelle commented on files/en-us/web/html/attributes/placeholder/index.md:
```diff
@@ -0,0 +1,52 @@
+
+The **`placeholder`** attribute is supported by the following input types: `{{HTMLElement("input/text", "text")}}`, `{{HTMLElement("input/search", "search")}}`, `{{HTMLElement("input/url", "url")}}`, `{{HTMLElement("input/tel", "tel")}}`, `{{HTMLElement("input/email", "email")}}`, and `{{HTMLElement("input/password", "password")}}`. It is also supported by the `{{HTMLElement("textarea")}}` elemen ...
+
+> **Note:** For a longer hint, or other advisory text, the title attribute is more appropriate.
```
title attributes are not accessible and not a replacement.They should be using a label.

>  These mechanisms are very similar but subtly different: the hint given by the control's [label](https://html.spec.whatwg.org/#the-label-element) is shown at all times; the short hint given in the [placeholder](https://html.spec.whatwg.org/#attr-input-placeholder) attribute is shown before the user enters a value; and the hint in the [title](https://html.spec.whatwg.org/#attr-title) attribute is shown when the user requests further help.

estelle reviewed (changes requested):
Take a look at https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-placeholder which explains the purpose of placeholders well. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#placeholders_are_not_accessible is also better.

Arisch24 commented on files/en-us/web/html/attributes/placeholder/index.md:
```diff
@@ -0,0 +1,52 @@
+## See also
+
+- [HTML attribute: title](/en-US/docs/Web/HTML/Global_attributes/title)
+- [CSS selector `::placeholder`](/en-US/docs/Web/CSS/::placeholder)
```
Can I know why is this preferred?

Arisch24 commented:
Thank you for the suggestions @estelle. I will proceed to fix them.

Arisch24 committed:
Language and grammar and format fixes (#28511)

Fix HTML semantics, EmbedLiveSample, link formatting and added further explanations regarding placeholder attributes

estelle commented on files/en-us/web/html/attributes/placeholder/index.md:
```diff
@@ -0,0 +1,54 @@
+
+The **`placeholder`** attribute is supported by the following input types: `{{HTMLElement("input/text", "text")}}`, `{{HTMLElement("input/search", "search")}}`, `{{HTMLElement("input/url", "url")}}`, `{{HTMLElement("input/tel", "tel")}}`, `{{HTMLElement("input/email", "email")}}`, and `{{HTMLElement("input/password", "password")}}`. It is also supported by the `{{HTMLElement("textarea")}}` elemen ...
+
+> **Note:** The attribute, if specified, must have a value that contains no line feeds (LF) or carriage returns (CR) characters. If the value contains either LF or CR characters, the value will be clipped.
+
+> **Note:** For a longer hint, or other advisory text, the {{HTMLElement("small")}} should be used.
+
+> **Warning:** Placeholders should only be used to show an example of the type of data that should be entered into a form; they don't replace a proper label as they are not semantically useful. This harms many users who are either older or people with disabilities. For example, older users may not understand that the placeholder text disappears when they start typing, and users with cognitive imp ...
```
```suggestion
The `placeholder` attribute can't include any line feeds (LF) or carriage returns (CR). If either is included in the value, the placeholder text will be clipped.

## Accessibility concerns

Placeholders should only be used to show an example of the type of data that should be entered into a form; never as a replacement for a `<label>` element; doing so harms accessibility and user experience. 

The `<label>` text is visually and programmatically associated with its corresponding form control. Screen readers don't announce placeholder content by default, but they do announce label content; it's the label that informs assistive technology users what data should be entered in the control. Labels also improve user experience for users of pointing devices: When a user clicks, touches, or taps a `<label>`, focus is moved to the the label's associated form control.

Placeholders can not be relied upon as a replacement for a label even for those not relying on assistive technology.  Placeholder text is displayed at lower color contrast than the default form control text. This is by design, as you don't want users to be confused by what is placeholder text versus what is a filled-in form field. However, this lack of contrast can cause issues for low-vision users. Additionally, placeholder text disappears from form fields when users start entering text. If the placeholder text contains instructional information or examples that disappear, it can be confusing to users with cognitive issues and can make the form inaccessible if the placeholder contained the label.
```

estelle replied:
To include the small text comment, we should link to the resource explaining how to write effective longer hints and advisory text

Arisch24 committed:
More explanation and minor fixes #28511

estelle reviewed (approved):
Yay! We did it! Thank you for all your work.

For future PR reviews, let us know when you are ready for a re-review.

To send us a notification, click on the icon next to the reviewer's name in the top right of the page (widescreen) or after the last comment (narrow screen). (The id of the section is #reviewers-select-menu if you can't find it). When the icon is clicked, it changes from an icon with two arrows to a brown dot and the reviewer gets a notification.

estelle committed:
Apply suggestions from code review

estelle committed:
Merge branch 'main' into arisch24-placeholder-attr

Arisch24 commented:
Thank for your support and guidance @estelle. Yeah next time I'll be sure to send the notification when I'm ready.
